### PR TITLE
add length type parameter to StepRangeLen

### LINKED ISSUE #41517: Julia v1.7: Multiplication broken with `BigInt` ranges
In Julia v1.7-beta:
```julia
julia> 2 * (1:big(100)^100)
ERROR: InexactError: Int64(100000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000)
Stacktrace:
 [1] Type
   @ ./gmp.jl:363 [inlined]
 [2] convert
   @ ./number.jl:7 [inlined]
 [3] StepRangeLen{BigInt, BigInt, BigInt}(ref::BigInt, step::BigInt, len::BigInt, offset::Int64)
   @ Base ./range.jl:441
 [4] StepRangeLen (repeats 2 times)
   @ ./range.jl:445 [inlined]
 [5] broadcasted(#unused#::Base.Broadcast.DefaultArrayStyle{1}, #unused#::typeof(*), x::Int64, r::UnitRange{BigInt})
   @ Base.Broadcast ./broadcast.jl:1157
 [6] broadcasted
   @ ./broadcast.jl:1341 [inlined]
 [7] broadcast_preserving_zero_d
   @ ./broadcast.jl:892 [inlined]
 [8] *(A::Int64, B::UnitRange{BigInt})
   @ Base ./arraymath.jl:52
 [9] top-level scope
   @ REPL[9]:1
```

Previously:
```julia
julia> 2 * (1:big(100)^100)
2:2:200000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
```

The issue is that `StepRangeLen` only allows `Int` lengths.

### PR DESCRIPTION
**Pull request by vtjnash**
Also be more careful about using additive identity instead of
multiplicative, and be more consistent about types in a few places.

Fixes #41517

### CODE REVIEW

File: base/range.jl [hunk 1]
```diff
@@ -426,26 +427,30 @@ value `r[1]`, but alternatively you can supply it as the value of
 with `TwicePrecision` this can be used to implement ranges that are
 free of roundoff error.
 """
-struct StepRangeLen{T,R,S} <: AbstractRange{T}
+struct StepRangeLen{T,R,S,L} <: AbstractRange{T}
```
Comment (sostock) [inline base/range.jl:430]:
```suggestion
struct StepRangeLen{T,R,S,L<:Integer} <: AbstractRange{T}
```
Or is there a reason to leave `L` unconstrained?

Comment (vtjnash) [reply base/range.jl:430]: 👍 it is limited by the (only) constructor, so probably makes sense to limit here too

File: base/twiceprecision.jl [hunk 2]
```diff
@@ -638,25 +648,26 @@ function range_start_stop_length(start::T, stop::T, len::Integer) where {T<:IEEE
...
         ref = T((1-t)*start + t*stop)
         step = imin-1 < len-imin ? (ref-start)/(imin-1) : (stop-ref)/(len-imin)
     elseif imin <= 1
-        imin = 1
+        imin = oneunit(typeof(len))
         ref = start
         step = (Δ/(len-1))*Δfac
     else
-        imin = Int(len)
+        imin = len
```
Comment (JeffBezanson) [inline base/twiceprecision.jl:670]: The `2` and `1` on the call to `steprangelen_hp` below might also need their types adjusted.

File: base/range.jl [hunk 3]
```diff
@@ -803,7 +809,13 @@ copy(r::AbstractRange) = r
 
 ## iteration
 
-function iterate(r::Union{LinRange,StepRangeLen}, i::Int=1)
+function iterate(r::StepRangeLen, i::Integer=1)
```
Comment (JeffBezanson) [inline base/range.jl:812]:
```suggestion
function iterate(r::StepRangeLen, i::Integer=oneunit(length(r)))
```

File: base/range.jl [hunk 4]
```diff
@@ -426,26 +427,30 @@ value `r[1]`, but alternatively you can supply it as the value of
...
+    function StepRangeLen{T,R,S,L}(ref::R, step::S, len::Integer, offset::Integer = 1) where {T,R,S,L}
         if T <: Integer && !isinteger(ref + step)
             throw(ArgumentError("StepRangeLen{<:Integer} cannot have non-integer step"))
         end
+        len = convert(L, len)
         len >= 0 || throw(ArgumentError("length cannot be negative, got $len"))
-        1 <= offset <= max(1,len) || throw(ArgumentError("StepRangeLen: offset must be in [1,$len], got $offset"))
-        new(ref, step, len, offset)
+        offset = convert(L, offset)
+        1 <= offset <= max(1, len) || throw(ArgumentError("StepRangeLen: offset must be in [1,$len], got $offset"))
```
[the remark this answers is not in the document]
Comment (vtjnash) [reply base/range.jl]: I think I will leave that to someone else. This has grown already into a monstrosity (and I would like to see codegen do that transform automatically instead)

**Comment by vtjnash** [comment]
@JeffBezanson could you re-review? this now transforms LinRange too, for the sake of completeness

File: base/range.jl [hunk 5]
```diff
@@ -415,8 +415,9 @@ oneto(r) = OneTo(r)
 ## Step ranges parameterized by length
 
 """
-    StepRangeLen{T,R,S}(ref::R, step::S, len, [offset=1]) where {T,R,S}
-    StepRangeLen(       ref::R, step::S, len, [offset=1]) where {  R,S}
+    StepRangeLen(         ref::R, step::S, len, [offset=1]) where {  R,S}
+    StepRangeLen{T,R,S}(  ref::R, step::S, len, [offset=1]) where {T,R,S}
+    StepRangeLen{T,R,S,L}(ref::R, step::S, len, [offset=1]) where {T,R,S,L}
 
 A range `r` where `r[i]` produces values of type `T` (in the second
```
Comment (JeffBezanson) [inline base/range.jl:422]:
```suggestion
A range `r` where `r[i]` produces values of type `T` (in the first
```

File: base/broadcast.jl [hunk 6]
```diff
@@ -1121,19 +1121,20 @@ end
...
 broadcasted(::DefaultArrayStyle{1}, ::typeof(-), r::StepRangeLen) = StepRangeLen(-r.ref, -r.step, length(r), r.offset)
 broadcasted(::DefaultArrayStyle{1}, ::typeof(-), r::LinRange) = LinRange(-r.start, -r.stop, length(r))
 
-broadcasted(::DefaultArrayStyle{1}, ::typeof(+), x::Real, r::AbstractUnitRange) = range(x + first(r), length=length(r))
-broadcasted(::DefaultArrayStyle{1}, ::typeof(+), r::AbstractUnitRange, x::Real) = range(first(r) + x, length=length(r))
 # For #18336 we need to prevent promotion of the step type:
 broadcasted(::DefaultArrayStyle{1}, ::typeof(+), r::AbstractRange, x::Number) = range(first(r) + x, step=step(r), length=length(r))
 broadcasted(::DefaultArrayStyle{1}, ::typeof(+), x::Number, r::AbstractRange) = range(x + first(r), step=step(r), length=length(r))
+broadcasted(::DefaultArrayStyle{1}, ::typeof(+), r::OrdinalRange, x::Real) = range(first(r) + x, last(r) + x, step=step(r))
+broadcasted(::DefaultArrayStyle{1}, ::typeof(+), x::Real, r::Real) = range(x + first(r), x + last(r), step=step(r))
```
Comment (GunnarFarneback) [inline base/broadcast.jl:1135]: This line should supposedly have `r::OrdinalRange`, but cf #42291.

**Pull request merged**